### ISSUE TITLE
not(throwA) doesn't work or wrong usage?

### ISSUE THREAD

**Issue by britishdan** [etorreborre/specs2#395]
Is there a reason this test is green?

```scala
class ThrowTest extends Specification {
  class MyClass {
    def doSmth(): Unit = throw new IllegalArgumentException
  }

  "not(throwA)" should {
    "succeed if another exception was thrown" >> {
      val myClass = new MyClass
      myClass.doSmth() must not(throwA[IllegalStateException])
    }
  }
}
```

**Comment by britishdan** [comment]
FYI @dkomanov

**Comment by etorreborre** [comment]
This test is green because your code is throwing an `IllegalArgumentException` and not an `IllegalStateException`. Why would you expect something else?

**Comment by britishdan** [comment]
As written here:
http://www.komanov.com/#!Do-not-use-notthrowA-matcher/yvypt/55acfa6c0cf25466c2a49ed0
the matcher hides other exceptions.


On Tuesday, July 21, 2015, Eric Torreborre <notifications@github.com> wrote:

> This test is green because your code is throwing an
> IllegalArgumentException and not an IllegalStateException. Why would you
> expect something else?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/etorreborre/specs2/issues/395#issuecomment-123074665>.
>


-- 

Thanks,
*Daniel Coldham*
Server whisperer

Cell: +972 (52) 444-2586
40 Hanamal street, Tel Aviv, Israel

<http://www.wix.com>

"Everything not saved will be lost." -Nintendo "Quit Screen" message

**Comment by etorreborre** [comment]
I see what you mean now. In that case you could almost use no matcher because you are saying that you want no exceptions and particularly not an `IllegalStateException`, which you would get as errors in your test if you don't use any matcher. 

The "ideal" `not(throw)` matcher would be something like `failOn[IllegalStateException]` which would create a failure if that exception is caught and re-throw other exceptions
```
def failOn[T <: Throwable, R](implicit c: ClassTag[T]): Matcher[R] = (r: R) => 
   try r
   catch {
     case t if c.runtimeClass isAssignableFrom t.getClass =>(false, s"c.runtimeClass.getName must not be thrown")
     case other => throw other
   }
```

**Comment by britishdan** [comment]
Sounds good to me.

**Comment by dkomanov** [comment]
Eric, what do you think, this failOn should be in specs2 or not?

**Comment by etorreborre** [comment]
@dkomanov @britishdan I don't like exceptions but swallowed exceptions are even worse. I fixed this issue by overriding the general `not` behavior for matchers in the case of exception matchers:
```
 negated matcher: when we specify not(throw[A])
   + it must return an Error if B is thrown
   + it must return a Failure if A is thrown
   + it must return a Success if no exception is thrown
```

This is available in `3.6.2-20150722230651-341faec`.

**Comment by dkomanov** [comment]
Thanks!
